File: telegram_upload_watcher/queue.py
from __future__ import annotations

import json
import logging
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


STATUS_QUEUED = "queued"
STATUS_SENDING = "sending"
STATUS_SENT = "sent"
STATUS_FAILED = "failed"

PENDING_STATUSES = {STATUS_QUEUED, STATUS_FAILED}

QUEUE_META_TYPE = "queue_meta"
QUEUE_META_VERSION = 1


@dataclass
class QueueItem:
    id: str
    source_type: str
    source_path: str
    source_fingerprint: str
    path: str
    inner_path: str | None
    size: int
    mtime_ns: int | None
    crc: int | None
    fingerprint: str
    status: str
    enqueued_at: str
    updated_at: str
    send_type: str = "image"
    error: str | None = None


def _utc_now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _normalize_meta_value(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: _normalize_meta_value(val) for key, val in value.items()}
    if isinstance(value, list):
        normalized = [_normalize_meta_value(item) for item in value]
        try:
            return sorted(normalized)
        except TypeError:
            return normalized
    return value


def _normalize_meta(meta: dict[str, Any]) -> dict[str, Any]:
    return _normalize_meta_value(meta)


def _is_meta(payload: dict[str, Any]) -> bool:
    return (
        payload.get("type") == QUEUE_META_TYPE
        and payload.get("version") == QUEUE_META_VERSION
        and isinstance(payload.get("params"), dict)
    )


def _meta_params(payload: dict[str, Any]) -> dict[str, Any]:
    params = payload.get("params")
    if not isinstance(params, dict):
        return {}
    return _normalize_meta_value(params)


def _meta_matches(expected: dict[str, Any], actual: dict[str, Any]) -> bool:
    return _meta_params(expected) == _meta_params(actual)


def build_fingerprint(
    source_type: str,
    path: str,
    inner_path: str | None,
    size: int,
    mtime_ns: int | None,
    crc: int | None,
) -> str:
    parts = [source_type, path]
    if inner_path:
        parts.append(inner_path)
    parts.append(str(size))
    if mtime_ns is not None:
        parts.append(str(mtime_ns))
    if crc is not None:
        parts.append(str(crc))
    return "|".join(parts)


def build_source_fingerprint(
    source_path: str, size: int, mtime_ns: int | None
) -> str:
    parts = [source_path, str(size)]
    if mtime_ns is not None:
        parts.append(str(mtime_ns))
    return "|".join(parts)


class JsonlQueue:
    def __init__(self, path: Path, meta: dict[str, Any] | None = None) -> None:
        self.path = path
        self.items: dict[str, QueueItem] = {}
        self.fingerprint_index: dict[str, str] = {}
        self.source_index: set[str] = set()
        self.meta = _normalize_meta(meta) if meta is not None else None
        self._meta_checked = False
        self._meta_found = False
        self._load()
        self._ensure_meta()

    def _load(self) -> None:
        if not self.path.exists():
            return
        meta_checked = False
        meta_found = False
        try:
            with self.path.open("r", encoding="utf-8") as handle:
                for line in handle:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        payload = json.loads(line)
                    except json.JSONDecodeError:
                        logging.warning("Skipping invalid queue line")
                        if self.meta is not None and not meta_checked:
                            raise ValueError(
                                "Queue metadata missing or invalid. Use a different --queue-file."
                            )
                        continue
                    if not meta_checked:
                        meta_checked = True
                        if isinstance(payload, dict) and _is_meta(payload):
                            meta_found = True
                            if self.meta is not None and not _meta_matches(
                                self.meta, payload
                            ):
                                raise ValueError(
                                    "Queue metadata does not match current run parameters."
                                )
                            continue
                        if self.meta is not None:
                            raise ValueError(
                                "Queue metadata missing. Use a different --queue-file."
                            )
                    item_id = payload.get("id")
                    if not item_id:
                        continue
                    try:
                        item = QueueItem(**payload)
                    except TypeError:
                        logging.warning("Skipping incompatible queue entry")
                        continue
                    self.items[item_id] = item
        except FileNotFoundError:
            return

        self._meta_checked = meta_checked
        self._meta_found = meta_found
        self._rebuild_indexes()

    def _ensure_meta(self) -> None:
        if self.meta is None:
            return
        if self._meta_found or self._meta_checked:
            return
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self.path.open("w", encoding="utf-8") as handle:
            handle.write(
                json.dumps(self.meta, ensure_ascii=True, sort_keys=True) + "\n"
            )
        self._meta_checked = True
        self._meta_found = True

    def _rebuild_indexes(self) -> None:
        self.fingerprint_index.clear()
        self.source_index.clear()
        for item_id, item in self.items.items():
            self.fingerprint_index[item.fingerprint] = item_id
            source_key = f"{item.source_type}:{item.source_fingerprint}"
            self.source_index.add(source_key)

    def _append(self, item: QueueItem) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self.path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(item.__dict__, ensure_ascii=True) + "\n")

    def has_fingerprint(self, fingerprint: str) -> bool:
        return fingerprint in self.fingerprint_index

    def has_source_fingerprint(self, source_type: str, source_fingerprint: str) -> bool:
        return f"{source_type}:{source_fingerprint}" in self.source_index

    def enqueue_item(
        self,
        *,
        source_type: str,
        source_path: str,
        source_fingerprint: str,
        path: str,
        inner_path: str | None,
        size: int,
        mtime_ns: int | None,
        crc: int | None = None,
        send_type: str = "image",
    ) -> QueueItem | None:
        fingerprint = build_fingerprint(
            source_type, path, inner_path, size, mtime_ns, crc
        )
        if fingerprint in self.fingerprint_index:
            return None

        now = _utc_now()
        item = QueueItem(
            id=uuid.uuid4().hex,
            source_type=source_type,
            source_path=source_path,
            source_fingerprint=source_fingerprint,
            path=path,
            inner_path=inner_path,
            size=size,
            mtime_ns=mtime_ns,
            crc=crc,
            send_type=send_type,
            fingerprint=fingerprint,
            status=STATUS_QUEUED,
            enqueued_at=now,
            updated_at=now,
            error=None,
        )
        self.items[item.id] = item
        self.fingerprint_index[item.fingerprint] = item.id
        self.source_index.add(f"{source_type}:{source_fingerprint}")
        self._append(item)
        return item

    def update_status(self, item_id: str, status: str, error: str | None = None) -> None:
        item = self.items.get(item_id)
        if not item:
            raise KeyError(f"Queue item not found: {item_id}")
        item.status = status
        item.updated_at = _utc_now()
        item.error = error
        self._append(item)

    def get_pending(self, limit: int | None = None) -> list[QueueItem]:
        pending = [
            item
            for item in self.items.values()
            if item.status in PENDING_STATUSES
        ]
        pending.sort(key=lambda entry: entry.enqueued_at)
        if limit is not None:
            return pending[:limit]
        return pending

    def stats(self) -> dict[str, Any]:
        counts = {
            STATUS_QUEUED: 0,
            STATUS_SENDING: 0,
            STATUS_SENT: 0,
            STATUS_FAILED: 0,
        }
        for item in self.items.values():
            if item.status in counts:
                counts[item.status] += 1
        return counts

File: go/internal/queue/queue.go
package queue

import (
	"bufio"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

const (
	StatusQueued  = "queued"
	StatusSending = "sending"
	StatusSent    = "sent"
	StatusFailed  = "failed"

	MetaType    = "queue_meta"
	MetaVersion = 1
)

var pendingStatuses = map[string]bool{
	StatusQueued: true,
	StatusFailed: true,
}

type Meta struct {
	Type    string     `json:"type"`
	Version int        `json:"version"`
	Params  MetaParams `json:"params"`
}

type WatchDirs []string

func (w *WatchDirs) UnmarshalJSON(data []byte) error {
	if len(data) == 0 || string(data) == "null" {
		*w = nil
		return nil
	}
	if data[0] == '"' {
		var value string
		if err := json.Unmarshal(data, &value); err != nil {
			return err
		}
		value = strings.TrimSpace(value)
		if value == "" {
			*w = nil
			return nil
		}
		*w = WatchDirs{value}
		return nil
	}
	var values []string
	if err := json.Unmarshal(data, &values); err != nil {
		return err
	}
	cleaned := make([]string, 0, len(values))
	for _, value := range values {
		value = strings.TrimSpace(value)
		if value != "" {
			cleaned = append(cleaned, value)
		}
	}
	*w = WatchDirs(cleaned)
	return nil
}

type MetaParams struct {
	Command   string    `json:"command"`
	WatchDir  WatchDirs `json:"watch_dir"`
	Recursive bool      `json:"recursive"`
	ChatID    string    `json:"chat_id"`
	TopicID   *int      `json:"topic_id,omitempty"`
	WithImage bool      `json:"with_image"`
	WithVideo bool      `json:"with_video"`
	WithAudio bool      `json:"with_audio"`
	WithAll   bool      `json:"with_all"`
	Include   []string  `json:"include,omitempty"`
	Exclude   []string  `json:"exclude,omitempty"`
}

type Item struct {
	ID                string  `json:"id"`
	SourceType        string  `json:"source_type"`
	SourcePath        string  `json:"source_path"`
	SourceFingerprint string  `json:"source_fingerprint"`
	Path              string  `json:"path"`
	InnerPath         *string `json:"inner_path,omitempty"`
	Size              int64   `json:"size"`
	MTimeNS           *int64  `json:"mtime_ns,omitempty"`
	CRC               *uint32 `json:"crc,omitempty"`
	SendType          string  `json:"send_type,omitempty"`
	Fingerprint       string  `json:"fingerprint"`
	Status            string  `json:"status"`
	EnqueuedAt        string  `json:"enqueued_at"`
	UpdatedAt         string  `json:"updated_at"`
	Error             *string `json:"error,omitempty"`
}

type Queue struct {
	path             string
	mu               sync.Mutex
	items            map[string]*Item
	fingerprintIndex map[string]string
	sourceIndex      map[string]struct{}
	appendCh         chan *Item
	closeCh          chan struct{}
	meta             *Meta
	metaChecked      bool
	metaFound        bool
}

func New(path string, meta *Meta) (*Queue, error) {
	q := &Queue{
		path:             path,
		items:            map[string]*Item{},
		fingerprintIndex: map[string]string{},
		sourceIndex:      map[string]struct{}{},
		appendCh:         make(chan *Item, 4096),
		closeCh:          make(chan struct{}),
		meta:             normalizeMeta(meta),
	}
	if err := q.load(); err != nil {
		return nil, err
	}
	if q.meta != nil && !q.metaChecked {
		if err := q.writeMeta(); err != nil {
			return nil, err
		}
	}
	go q.writerLoop()
	return q, nil
}

func buildID() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return hex.EncodeToString(buf), nil
}

func nowUTC() string {
	return time.Now().UTC().Format(time.RFC3339Nano)
}

func normalizeMeta(meta *Meta) *Meta {
	if meta == nil {
		return nil
	}
	copyMeta := *meta
	if copyMeta.Type == "" {
		copyMeta.Type = MetaType
	}
	if copyMeta.Version == 0 {
		copyMeta.Version = MetaVersion
	}
	copyMeta.Params.WatchDir = normalizeWatchDirs(meta.Params.WatchDir)
	copyMeta.Params.Include = append([]string{}, meta.Params.Include...)
	copyMeta.Params.Exclude = append([]string{}, meta.Params.Exclude...)
	sort.Strings(copyMeta.Params.Include)
	sort.Strings(copyMeta.Params.Exclude)
	return &copyMeta
}

func normalizeWatchDirs(dirs WatchDirs) WatchDirs {
	if len(dirs) == 0 {
		return nil
	}
	cleaned := make([]string, 0, len(dirs))
	for _, dir := range dirs {
		dir = strings.TrimSpace(dir)
		if dir != "" {
			cleaned = append(cleaned, dir)
		}
	}
	sort.Strings(cleaned)
	if len(cleaned) == 0 {
		return nil
	}
	return WatchDirs(cleaned)
}

func parseMeta(line string) (*Meta, bool, error) {
	var meta Meta
	if err := json.Unmarshal([]byte(line), &meta); err != nil {
		return nil, false, err
	}
	if meta.Type != MetaType || meta.Version != MetaVersion {
		return nil, false, nil
	}
	return normalizeMeta(&meta), true, nil
}

func metaMatches(expected *Meta, actual *Meta) bool {
	if expected == nil || actual == nil {
		return false
	}
	return reflect.DeepEqual(normalizeMeta(expected), normalizeMeta(actual))
}

func BuildFingerprint(sourceType, path string, innerPath *string, size int64, mtimeNS *int64, crc *uint32) string {
	parts := []string{sourceType, path, strconv64(size)}
	if innerPath != nil && *innerPath != "" {
		parts = append(parts, *innerPath)
	}
	if mtimeNS != nil {
		parts = append(parts, strconv64(*mtimeNS))
	}
	if crc != nil {
		parts = append(parts, strconv64(int64(*crc)))
	}
	return strings.Join(parts, "|")
}

func BuildSourceFingerprint(path string, size int64, mtimeNS *int64) string {
	parts := []string{path, strconv64(size)}
	if mtimeNS != nil {
		parts = append(parts, strconv64(*mtimeNS))
	}
	return strings.Join(parts, "|")
}

func strconv64(value int64) string {
	return strconv.FormatInt(value, 10)
}

func (q *Queue) load() error {
	file, err := os.Open(q.path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	buffer := make([]byte, 0, 1024*1024)
	scanner.Buffer(buffer, 1024*1024)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		if !q.metaChecked {
			q.metaChecked = true
			meta, ok, err := parseMeta(line)
			if err != nil {
				if q.meta != nil {
					return errors.New("queue metadata missing or invalid. Use a different --queue-file")
				}
				continue
			}
			if ok {
				q.metaFound = true
				if q.meta != nil && !metaMatches(q.meta, meta) {
					return errors.New("queue metadata does not match current run parameters")
				}
				continue
			}
			if q.meta != nil {
				return errors.New("queue metadata missing. Use a different --queue-file")
			}
		}
		var item Item
		if err := json.Unmarshal([]byte(line), &item); err != nil {
			continue
		}
		if item.ID == "" {
			continue
		}
		q.items[item.ID] = &item
	}
	q.rebuildIndexes()
	return scanner.Err()
}

func (q *Queue) rebuildIndexes() {
	q.fingerprintIndex = map[string]string{}
	q.sourceIndex = map[string]struct{}{}
	for id, item := range q.items {
		q.fingerprintIndex[item.Fingerprint] = id
		key := item.SourceType + ":" + item.SourceFingerprint
		q.sourceIndex[key] = struct{}{}
	}
}

func (q *Queue) writeMeta() error {
	if q.meta == nil {
		return nil
	}
	q.meta = normalizeMeta(q.meta)
	if err := os.MkdirAll(filepath.Dir(q.path), 0o755); err != nil {
		return err
	}
	file, err := os.OpenFile(q.path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o644)
	if err != nil {
		return err
	}
	defer file.Close()
	data, err := json.Marshal(q.meta)
	if err != nil {
		return err
	}
	if _, err := file.Write(append(data, '\n')); err != nil {
		return err
	}
	q.metaChecked = true
	q.metaFound = true
	return nil
}

func (q *Queue) writerLoop() {
	file, err := os.OpenFile(q.path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return
	}
	defer file.Close()
	writer := bufio.NewWriter(file)
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	batch := []*Item{}
	flush := func() {
		if len(batch) == 0 {
			return
		}
		for _, item := range batch {
			data, err := json.Marshal(item)
			if err != nil {
				continue
			}
			writer.Write(data)
			writer.WriteByte('\n')
		}
		writer.Flush()
		batch = batch[:0]
	}

	for {
		select {
		case item := <-q.appendCh:
			batch = append(batch, item)
			if len(batch) >= 128 {
				flush()
			}
		case <-ticker.C:
			flush()
		case <-q.closeCh:
			flush()
			return
		}
	}
}

func (q *Queue) Close() {
	close(q.closeCh)
}

func (q *Queue) HasFingerprint(fingerprint string) bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	_, ok := q.fingerprintIndex[fingerprint]
	return ok
}

func (q *Queue) HasSourceFingerprint(sourceType, sourceFingerprint string) bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	_, ok := q.sourceIndex[sourceType+":"+sourceFingerprint]
	return ok
}

func (q *Queue) Enqueue(item Item) (*Item, error) {
	q.mu.Lock()
	defer q.mu.Unlock()

	if item.Fingerprint == "" {
		return nil, errors.New("missing fingerprint")
	}
	if _, exists := q.fingerprintIndex[item.Fingerprint]; exists {
		return nil, nil
	}

	id, err := buildID()
	if err != nil {
		return nil, err
	}
	now := nowUTC()
	item.ID = id
	item.Status = StatusQueued
	item.EnqueuedAt = now
	item.UpdatedAt = now

	q.items[item.ID] = &item
	q.fingerprintIndex[item.Fingerprint] = item.ID
	q.sourceIndex[item.SourceType+":"+item.SourceFingerprint] = struct{}{}

	q.appendCh <- &item
	return &item, nil
}

func (q *Queue) UpdateStatus(id string, status string, errMsg *string) error {
	q.mu.Lock()
	defer q.mu.Unlock()
	item, ok := q.items[id]
	if !ok {
		return errors.New("queue item not found")
	}
	item.Status = status
	item.UpdatedAt = nowUTC()
	item.Error = errMsg
	q.appendCh <- item
	return nil
}

func (q *Queue) Pending(limit int) []*Item {
	q.mu.Lock()
	defer q.mu.Unlock()
	pending := []*Item{}
	for _, item := range q.items {
		if pendingStatuses[item.Status] {
			pending = append(pending, item)
		}
	}
	sort.Slice(pending, func(i, j int) bool {
		return pending[i].EnqueuedAt < pending[j].EnqueuedAt
	})
	if limit > 0 && len(pending) > limit {
		return pending[:limit]
	}
	return pending
}

func (q *Queue) Stats() map[string]int {
	q.mu.Lock()
	defer q.mu.Unlock()
	counts := map[string]int{
		StatusQueued:  0,
		StatusSending: 0,
		StatusSent:    0,
		StatusFailed:  0,
	}
	for _, item := range q.items {
		if _, ok := counts[item.Status]; ok {
			counts[item.Status]++
		}
	}
	return counts
}
